cmd/clusterctl/client/repository/goproxy.go

```diff
@@ -0,0 +1,163 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/blang/semver"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type goproxyClient struct {
+	scheme string
+	host   string
+}
+
+func newGoproxyClient(scheme, host string) *goproxyClient {
+	return &goproxyClient{
+		scheme: scheme,
+		host:   host,
+	}
+}
+
+func (g *goproxyClient) getVersions(owner, repository string) ([]string, error) {
+	// build the naive path
+	var gomodulePath string
+	switch owner {
+	case "kubernetes-sigs":
+		gomodulePath = path.Join("sigs.k8s.io", repository)
+	case "kubernetes":
+		gomodulePath = path.Join("k8s.io", repository)
+	default:
+		gomodulePath = path.Join("github.com", owner, repository)
+	}
+
+	rawURL := url.URL{
+		Scheme: g.scheme,
+		Host:   g.host,
+		Path:   path.Join(gomodulePath, "@v", "/list"),
+	}
+
+	var rawResponse []byte
+	var retryError error
+	_ = wait.PollImmediate(retryableOperationInterval, retryableOperationTimeout, func() (bool, error) {
+		req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, rawURL.String(), http.NoBody)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to get versions for go module %q via goproxy %q: failed to create request", gomodulePath, g.host)
+		}
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			retryError = errors.Wrapf(err, "failed to get versions for go module %q via goproxy %q", gomodulePath, g.host)
+			return false, nil
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != 200 {
+			retryError = errors.Errorf("Request status code %d: %s", resp.StatusCode, resp.Status)
+			return false, nil
+		}
+
+		rawResponse, err = io.ReadAll(resp.Body)
+		if err != nil {
+			return false, errors.Wrap(err, "error read goproxy response body")
+		}
+		return true, nil
+	})
+	if retryError != nil {
+		return nil, retryError
+	}
+
+	parsedVersions := semver.Versions{}
+	for _, s := range strings.Split(string(rawResponse), "\n") {
+		if s == "" {
+			continue
+		}
+		parsedVersion, err := semver.Parse(strings.TrimPrefix(s, "v"))
+		if err != nil {
+			// Discard releases with tags that are not a valid semantic versions (the user can point explicitly to such releases).
+			continue
+		}
+		parsedVersions = append(parsedVersions, parsedVersion)
+	}
+
+	sort.Sort(parsedVersions)
+
+	versions := []string{}
+	for _, v := range parsedVersions {
+		versions = append(versions, "v"+v.String())
+	}
+
+	return versions, nil
+}
+
+func getGoproxyHost(goproxy string) (string, string, error) {
+	// fallback default
+	if goproxy == "" {
+		return "https", "proxy.golang.org", nil
+	}
+
+	var goproxyHost, goproxyScheme string
+	// Use url from GOPROXY variable if set.
```

Comment:
this is slightly confusing because it refers to GOPROXY env var, but actually, we never read it in this func

Reply:
ack, was actual a leftover.